### DISCUSSION TITLE
Insert Client IP Address in Field

### DISCUSSION THREAD

**Discussion by DedsonDev** [pocketbase/pocketbase#3094] [Q&A] [answered]
I'm not sure if this is possible or not, but is there a way to get the client ip and insert it as a field in a collection?

If I have to use an event hook, how do I access the client ip?

**Comment by ganigeorgiev** [comment, marked as answer]
There is no such functionality out of the box but you can implement it on your own via the app hooks (either with Go or JS).
The main problem that you'll stumble on will be the IP extraction and not the persistence, because finding the real IP of the client may not be as easy as you may think (especially if you need it to be "trusted").

Here is a simple example that will insert the remote-address(!) via the `OnRecordBeforeCreateRequest` app hook (for more details how to extend PocketBase you can check the ["Use as framework" guide](https://pocketbase.io/docs/use-as-framework/)):

```go
// GO - https://pocketbase.io/docs/go-event-hooks/#onrecordbeforecreaterequest
app.OnRecordBeforeCreateRequest("yourCollection").Add(func(e *core.RecordCreateEvent) error {
    // this is the only address you can trust out of the box
    remoteAddr := e.HttpContext.Request().RemoteAddr

    e.Record.Set("ip", remoteAddr) // ip is a text field in the collection

    return nil
})
```

```js
// JS - https://pocketbase.io/docs/js-event-hooks/#onrecordbeforecreaterequest
onRecordBeforeCreateRequest((e) => {
    // this is the only address you can trust out of the box
    const remoteAddr = e.httpContext.request().remoteAddr

    e.record.set("ip", remoteAddr) // ip is a text field in the collection
}, "yourCollection")
```

Note that in the above we are using the `RemoteAddr` field. This is the only address that you can trust out of the box **but if you are behind a reverse proxy/gateway (eg. nginx, Cloudflare or even hosting on Fly.io) then this field will return most likely the address of the proxy server**.

In those cases the IP is usually set as a special header. For Cloudflare this header is usually named `CF-Connecting-IP`, for Fly.io it is `Fly-Client-IP`, etc. Most proxies also add the "real" client IP as part of the `X-Forwarded-For` header (it could contain more than one IP!) but you can't trust this header unless you are sure that your proxy overwrites/ignores any custom client sent value.

If you are OK with untrusted client IP extraction, you can replace the `remoteAddr` in the above examples with `e.HttpContext.RealIP()` (for Go) and `e.httpContext.realIP()` (for JS).

**Reply by DedsonDev** [reply]
Thank you for the detailed answer, this will work great for my purposes!
